frontend-react/docs/best-practices.md

```diff
@@ -0,0 +1,677 @@
+            const nameB = b.name.toUpperCase();
+
+            if (nameA < nameB) {
+                return -1;
+            }
+
+            if (nameA > nameB) {
+                return 1;
+            }
+
+            return 0;
+        });
+    }, [users, sortProperty]);
+    
+    return (
+        <ul>
+            { sortedUsers.map((user) => (
+                <li key={user.name}>{user.name}, {user.age}</li>
+            )) }
+        </ul>
+    )
+}
+```  
+
+Unless we need the reusable logic to tap into render functionality (or it'd be difficult to extricate otherwise), we can just turn this into a plain JavaScript function that can be exported/imported from anywhere (including inside a component or a hook):
+
+```tsx
+function sortUsers(users: User[], sortProperty: "age" | "name") {
+    if (sortProperty === "age") {
+        return users.sort((a, b) => a.age - b.age);
+    }
+
+    return users.sort((a, b) => {
+        const nameA = a.name.toUpperCase();
+        const nameB = b.name.toUpperCase();
+
+        if (nameA < nameB) {
+            return -1;
+        }
+
+        if (nameA > nameB) {
+            return 1;
+        }
+
+        return 0;
+    });
+}  
+
+// then just call this function in the component
+function UserList({ users, sortProperty }: { users: User[], sortProperty: "age" | "name" }) {
+    const sortedUsers = useMemo(() => {
+        return sortUsers(users, sortProperty);
+    }, [users, sortProperty]);
+
+    return (
+        <ul>
+            { sortedUsers.map((user) => (
+                <li key={user.name}>{user.name}, {user.age}</li>
+            )) }
+        </ul>
+    )
+}
+```  
+
+
+### Strive for single-purpose ("dumb") components and hooks
+
+> Why?
+>
+> This keeps components and hooks much more flexible in usage, which lends to the React ethos of composition.  Additionally, this breaks complex logic down for easier maintenance; ensuring that each component is focused around one purpose helps break up long logic chains for easier testability, debuggability, and upkeep, and it (usually) also helps keep the components and hooks smaller so they're more likely to be reusable across the codebase.
+
+The concept of "dumb" components has been around in the React ecosystem for a while: components whose sole purpose is to present and display data.  And although this concept was much easier to abide by with the [container/presenter pattern](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) before the introduction of hooks, keeping this clear separation of concerns is still important for scalability.
+
+Put simply, **dumb components are presentational whereas smart components are concerned with maintaining events and state.**  When writing a new component, it's important to question whether all the logic should exist in one place or if it would be more maintainable (and possibly readable) split out between different components or utility functions.
```

Comment:
might be worth replacing smart / dumb with something like stateful / stateless components or some other more apt descriptor?

Reply:
Hahah "smart" and "dumb" are pretty frequently used terms for components in front-end frameworks!  I've also heard "container" and "presentational" components, which could help to clarify.  Happy to change to that if preferred.

But "stateful" and "stateless" would be misnomers because that'd imply that no presentational component has state -- that's not true since there could be _component_ state through `useState`.  We're really talking about global or mediator state, but I'd argue that this'd still be confusing.

Reply:
That's totally fair, I was just trying to set up some meaningful dichotomy. Container and presentational sounds good to me!